### PR TITLE
Added ASLFeat extractor

### PR DESCRIPTION
**Pull request by travisdriver**

### CODE REVIEW

File: gtsfm/frontend/detector_descriptor/aslfeat.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        Args:
+            image: the input image.
+
+        Returns:
+            Detected keypoints, with length N <= max_keypoints.
+            Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
+        """
+        tf.reset_default_graph()
+
+        # Load model and image.
```
Comment (akshay-krishnan) [inline gtsfm/frontend/detector_descriptor/aslfeat.py:84]: is this run in inference mode? does TF compute gradients here?

**Review by akshay-krishnan — COMMENTED** [review]
Have you presented about this descriptor before? How useful is it?

**Comment by travisdriver** [comment]
> Have you presented about this descriptor before? How useful is it?

We found it to be the most effective feature for the AstriVision data: https://arxiv.org/pdf/2208.02053.pdf.

File: gtsfm/frontend/detector_descriptor/aslfeat.py (hunk 1 above)
Comment (travisdriver) [reply gtsfm/frontend/detector_descriptor/aslfeat.py:84]: It's run in inference mode.

File: gtsfm/frontend/detector_descriptor/aslfeat.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    def __init__(
+        self,
+        max_keypoints: int = 5000,
+        model_path: Path = DEFAULT_MODEL_PATH,
+        model_config: Dict[str, Any] = DEFAULT_CONFIG,
+    ) -> None:
+        """Instantiate parameters and hardware settings for D2-Net detector-descriptor.
+
+        We set the maximum number of keypoints, set the path to pre-trained weights, and determine whether
+        CUDA enabled devices can be utilized for inference.
```
[the remark this answers is not in the document]
Comment (travisdriver) [reply gtsfm/frontend/detector_descriptor/aslfeat.py:62]: Sorry, I use D2-Net as the template to create the class, and I forgot to change the description. For TensorFlow, you don't have to explicitly set the device to cuda as in PyTorch.

[the remark this answers is not in the document]
Comment (travisdriver) [reply gtsfm/frontend/detector_descriptor/aslfeat.py]: Updated docstrings

File: environment_linux.yml [hunk 3]
```diff
@@ -44,11 +44,12 @@ dependencies:
...
   - pip:
-    - open3d
-    - opencv-python>=4.5.4.58
-    - pydegensac
-    - colour
-    - pycolmap>=0.1.0
-    - trimesh[easy]
-    - gtsam==4.2a8
-    - pydot
+      - open3d
```
Comment (akshay-krishnan) [inline environment_linux.yml:47]: why the indentation change?

File: environment_linux_cpuonly.yml [hunk 4]
```diff
@@ -44,11 +44,12 @@ dependencies:
...
   - pip:
-    - open3d
-    - opencv-python>=4.5.4.58
-    - pydegensac
-    - colour
-    - pycolmap>=0.1.0
-    - trimesh[easy]
-    - gtsam==4.2a8
-    - pydot
+      - open3d
```
Comment (akshay-krishnan) [inline environment_linux_cpuonly.yml:47]: why the indentation change?

File: gtsfm/frontend/detector_descriptor/aslfeat.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+from typing import Tuple, Dict, Any
+import sys
+
+import numpy as np
+import tensorflow.compat.v1 as tf
+
+import gtsfm.utils.images as image_utils
+from gtsfm.common.image import Image
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.frontend.detector_descriptor.detector_descriptor_base import DetectorDescriptorBase
```
Comment (akshay-krishnan) [inline gtsfm/frontend/detector_descriptor/aslfeat.py:19]: do we do the from * imports first or the import * imports first? Looks like youve done both, "from imports" first for lines 9-11 and "import" imports first for 16-19.

File: gtsfm/frontend/detector_descriptor/aslfeat.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+sys.path.append("thirdparty/ASLFeat")
```
Comment (akshay-krishnan) [inline gtsfm/frontend/detector_descriptor/aslfeat.py:21]: do we need this? are the thirdparty modules not added without it?

File: gtsfm/utils/viz.py [hunk 7]
```diff
@@ -103,7 +110,7 @@ def plot_twoview_correspondences(
     corr_idxs_i1i2: np.ndarray,
     inlier_mask: Optional[np.ndarray] = None,
     dot_color: Optional[Tuple[int, int, int]] = None,
-    max_corrs: Optional[int] = 50,
+    max_corrs: Optional[int] = 5000,
```
Comment (akshay-krishnan) [inline gtsfm/utils/viz.py:113]: I think 5000 is too many, drawing so many is time consuming and makes the image cluttered. where have you found this useful?

File: thirdparty/ASLFeat/LICENSE [hunk 8]
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Comment (akshay-krishnan) [inline thirdparty/ASLFeat/LICENSE:1]: Could you update thirdparty/README.md?

**Comment by akshay-krishnan** [comment]
Is there a config that runs ASLFeat on the CI? 
Have you run it on other datasets? are there any others that benefit from it?

File: gtsfm/frontend/detector_descriptor/aslfeat.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            "nms_size": 3,
+            "eof_mask": 5,
+            "need_norm": True,
+            "use_peakiness": True,
+        },
+    }
+}
+
+
+class ASLFeatDetectorDescriptor(DetectorDescriptorBase):
```
Comment (akshay-krishnan) [inline gtsfm/frontend/detector_descriptor/aslfeat.py:53]: shouldnt we add some unit tests for this?

Comment (travisdriver) [reply gtsfm/frontend/detector_descriptor/aslfeat.py:53]: I was not able to get Tensorflow running on the CI

File: gtsfm/utils/viz.py (hunk 7 above)
Comment (travisdriver) [reply gtsfm/utils/viz.py:113]: It's difficult to assess the quality of the matches by only drawing a small fraction of them.  I could expose it to the config if you'd like?

File: gtsfm/frontend/detector_descriptor/aslfeat.py (hunk 6 above)
Comment (travisdriver) [reply gtsfm/frontend/detector_descriptor/aslfeat.py:21]: Yes, the ASLFeat library uses relative paths in the code. Although, I could fix these in the ASLFeat code since all of the files have been checked in?

File: environment_linux.yml (hunk 3 above)
Comment (travisdriver) [reply environment_linux.yml:47]: Done automatically by the Black formatter (see [CONTRIBUTING.md](https://github.com/borglab/gtsfm/blob/master/CONTRIBUTING.md))

File: environment_linux_cpuonly.yml (hunk 4 above)
Comment (travisdriver) [reply environment_linux_cpuonly.yml:47]: Done automatically by the Black formatter (see [CONTRIBUTING.md](https://github.com/borglab/gtsfm/blob/master/CONTRIBUTING.md))
